Add detailed reports

mbj opened the pull request:
Not intended for merge. Intended for review. PR will be re-done to upstream once flag free.

mbj commented on core/Test/Tasty/Ingredients/ConsoleReporter.hs:
```diff
@@ -135,6 +136,7 @@ buildTestOutput opts tree =
           when (not $ null rDesc) $
             (if resultSuccessful result then infoOk else infoFail) $
               printf "%s%s\n" (indent $ level + 1) (formatDesc (level+1) rDesc)
+          maybe (return ()) (\action -> action withConsoleFormat) (printResultDetails result)
```
Its likely possible to get a rid of the lambda here.

mbj replied:
I actually think this maybe should be removed in favor of: https://github.com/mbj/tasty/pull/1/commits/f4a22fbcbf2861adca73bf9f30488e47d64361c3#r337586802

dkubb replied:
@mbj could you replace this with:

```haskell
maybe (return ()) ($ withConsoleFormat) (printResultDetails result)
```

mbj replied:
I went with removing the maybe so this is redundant.

dkubb commented on core/tasty.cabal:
```diff
@@ -82,7 +82,7 @@ library
+  -- hs-source-dirs:
   default-language:    Haskell2010
   default-extensions:  CPP, ScopedTypeVariables, DeriveDataTypeable
   ghc-options: -Wall
```
Can this `Fix` commit be moved earlier?

mbj replied:
I wanted to keep it close to the `-Werror` addition, for easier upstream consumption. I'll try to sneak in not only our commit style but also our ordering.

mbj replied:
Fixed.

dkubb commented on core/Test/Tasty/ConsoleFormat.hs:
```diff
@@ -0,0 +1,24 @@
+
+data ConsoleFormat = ConsoleFormat
+  { consoleIntensity :: ConsoleIntensity
+  , colorIntensity   :: ColorIntensity
```
These keys can be sorted differently, but I realize there may be a semantic meaning behind the order.

mbj replied:
I went with the original upstream codes argument order. This way the format specifications below are a pure move.

So this derivation from our canonical form was intentional to minimize diff with upstream.

dkubb commented on core/Test/Tasty/Core.hs:
```diff
@@ -65,8 +68,21 @@ data Result = Result
+    <> shows (resultOutcome result) ", "
+    <> shows (resultDescription result) ", "
+    <> shows (resultShortDescription result) ", "
+    <> shows (resultTime result) ", "
```
You want a trailing `","` on this string? If not you could look at `intercalate` to simplify this a bit.

mbj replied:
The trailing "," is not intend, but it was intend to make it fuse nicely. Looking closer on how to apply `shows` correctly here.

mbj commented on core/Test/Tasty/Core.hs:
```diff
@@ -65,8 +68,21 @@ data Result = Result
     -- @FAIL@.
   , resultTime :: Time
     -- ^ How long it took to run the test, in seconds.
+  , printResultDetails :: Maybe ResultDetailsPrinter
```
I think this could also be `ResultDetailsPrinter` and we expose a `noopDetailsPrinter` to use in the defaults. This one would be implemented as `const $ return ()` (intentionally not using `pure` as this code base seems to prefer `return`).

snusnu commented on core/Test/Tasty/Core.hs:
```diff
@@ -65,8 +68,21 @@ data Result = Result
+  , printResultDetails :: Maybe ResultDetailsPrinter
+    -- ^ Optional detailed result printing action.
+    --
+    -- 'printDesultDetails' is used by providers that need to have fine grained
```
`s/printDesultDetails/printResultDetails/`

mbj commented on core/Test/Tasty/Core.hs:
```diff
@@ -65,8 +68,21 @@ data Result = Result
+
+instance Show Result where
+  show result
+    =  "Result {"
```
Still have to check before upstreaming if this `<>` and `shows` usage is efficient.

dkubb commented on core/Test/Tasty/Ingredients/ConsoleReporter.hs:
```diff
@@ -613,28 +616,33 @@ stringWidth = length
+output format = withConsoleFormat format . putStr
+
+withConsoleFormat
+  :: (?colors :: Bool)
```
What does `?colors` signify? I don't think I've seen this leading question mark syntax before.

mbj replied:
This is an implicit parameter, already present in the code. In this case it gets "bound by call graph".

See https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#implicit-parameters

dkubb commented on core/Test/Tasty/Providers.hs:
```diff
@@ -37,4 +40,13 @@ testFailed desc = Result
+  -> ResultDetailsPrinter -- ^ details printer
+  -> Result
+testFailedDetails desc printer = (testFailed desc)
+  { printResultDetails = Just printer }
```
I don't suppose we could sneak in `pure` or `return` here instead of `Just`?

mbj replied:
I wanted to have this as close to what upstream does as possible. Hence I used the `Just`. As I now removed the maybe the just is now redundant and we have one less thing to align for with upstream.

dkubb commented on core/Test/Tasty/Run.hs:
```diff
@@ -322,6 +323,7 @@ resolveDeps tests = checkCycles $ do
           , resultDescription = ""
           , resultShortDescription = "SKIP"
           , resultTime = 0
+          , printResultDetails = Nothing
```
Can we use `empty` or `mempty` here? And on line 160? I see no other cases where `Nothing` is used in this file.

mbj replied:
In this project: `empty` is never used `Nothing` a lot. Hence I chose to use `Nothing` also for consistency.

But as I removed the maybe around the printer, the decision for `Nothing` or another form is now redundant.

dkubb commented on core/Test/Tasty/Providers.hs:
```diff
@@ -37,4 +40,13 @@ testFailed desc = Result
   , resultDescription = desc
   , resultShortDescription = "FAIL"
   , resultTime = 0
+  , printResultDetails = Nothing
```
Instance of https://github.com/mbj/tasty/pull/1/commits/f4a22fbcbf2861adca73bf9f30488e47d64361c3#r337588780

mbj replied:
See: https://github.com/mbj/tasty/pull/1#discussion_r337602364

dkubb commented on core/Test/Tasty/Providers.hs:
```diff
@@ -26,6 +28,7 @@ testPassed desc = Result
   , resultDescription = desc
   , resultShortDescription = "OK"
   , resultTime = 0
+  , printResultDetails = Nothing
```
Instance of https://github.com/mbj/tasty/pull/1/commits/f4a22fbcbf2861adca73bf9f30488e47d64361c3#r337588780

mbj replied:
See: https://github.com/mbj/tasty/pull/1#discussion_r337602364

mbj committed:
Remove inline vim configuration

* Improves consistency with other files that do not share the inline
  configuration.
* Seeing these settings only in a single file appears to be a leftover.
* It should be the vim users choice about the presentation of the code.
  The vim user will have its own global config.

mbj committed:
Fix whitespace

mbj committed:
Add stack.yaml.lock

* This improves the reproducability of builds.

mbj committed:
Upgrade to LTS 14.11

mbj committed:
Add -Werror to tasty.cabal

mbj committed:
Add optional result details printer

* This details printer allows a provider full control about
  the report message.
* Its useful in case the description format would be loosing
  information due its default console formatting of green/red (depending
  on outcome).
* Allows to re-use the console formats used by the console formatter by
  providers that wish to use the new interface.
